src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -1383,16 +1386,94 @@ AddUnionAllSetOperationsToAttributeEquivalenceClass(AttributeEquivalenceClass **
 			continue;
 		}
 		int rtoffset = RangeTableOffsetCompat(root, appendRelInfo);
+		int childRelId = appendRelInfo->child_relid - rtoffset;
 
-		/* set the varno accordingly for this specific child */
-		varToBeAdded->varno = appendRelInfo->child_relid - rtoffset;
+		RangeTblEntry *rte = root->simple_rte_array[childRelId];
+		if (rte->inh)
+		{
+			/*
+			 * This code-path may require improvements. If a leaf of a UNION ALL
+			 * (e.g., an entry in appendRelList) itself is another UNION ALL
+			 * (e.g., rte->inh = true), the logic here might get into an infinite
+			 * recursion.
+			 *
+			 * The downside of "continue" here is that certain UNION ALL queries
+			 * that are safe to pushdown may not be pushed down.
+			 */
+			continue;
+		}
+		else if (rte->rtekind == RTE_RELATION)
+		{
+			Index partitionKeyIndex = 0;
+			List *translatedVars = TranslatedVarsForRteIdentity(GetRTEIdentity(rte));
+			Var *varOnUnionAllSubquery =
+				FindUnionAllVar(root, translatedVars, rte->relid, childRelId,
+								&partitionKeyIndex);
+			if (partitionKeyIndex == 0)
+			{
+				/* no partition key on the target list */
+				continue;
+			}
 
-		AddToAttributeEquivalenceClass(attributeEquivalenceClass, root,
-									   varToBeAdded);
+			if (unionQueryPartitionKeyIndex == 0)
+			{
+				/* the first partition key index we found */
+				unionQueryPartitionKeyIndex = partitionKeyIndex;
+			}
+			else if (unionQueryPartitionKeyIndex != partitionKeyIndex)
+			{
+				/*
+				 * Partition keys on the leaves of the UNION ALL queries on
+				 * different ordinal positions. We cannot pushdown, so skip.
+				 */
+				continue;
+			}
+
+			if (varOnUnionAllSubquery != NULL)
+			{
+				AddToAttributeEquivalenceClass(attributeEquivalenceClass, root,
+											   varOnUnionAllSubquery);
+			}
+		}
+		else
+		{
+			/* set the varno accordingly for this specific child */
+			varToBeAdded->varno = childRelId;
+
+			AddToAttributeEquivalenceClass(attributeEquivalenceClass, root,
+										   varToBeAdded);
+		}
 	}
 }
 
 
+/*
+ * TranslatedVarsForRteIdentity gets an rteIdentity and returns the
+ * translatedVars that belong to the range table relation. If no
+ * translatedVars found, the function returns NIL;
+ */
+static List *
+TranslatedVarsForRteIdentity(int rteIdentity)
+{
+	PlannerRestrictionContext *currentPlannerRestrictionContext =
```

Comment:
It is not the best approach to call `CurrentPlannerRestrictionContext ()` at this point. However, otherwise all the APIs `(starting from AddToAttributeEquivalenceClass())` need to change significantly and I wanted to get some feedback before doing that.

Reply:
> Or what about we move this method to distributed_planner ?

I think this is a good idea, and yes keeping `CurrentPlannerRestrictionContext ` static is better

Reply:
> umm, can we avoid making CurrentPlannerRestrictionContext public? I think we could have a method like CurrentRelationRestrictionList that returns the restriction list, which is what we need here.

I don't see much difference if we move this function to distributed_planner, what do you think?

Reply:
if we move to distributed_planner then we don't need to have `CurrentRelationRestrictionList`